### PR TITLE
Add chat subsystem to prevent client crash on high loads

### PR DESCRIPTION
**Pull request by Carbonhell**
## About The Pull Request
<details>
<summary>
	Fixes client crashes when spammed with a lot of messages due to goonchat (Ports https://github.com/tgstation/tgstation/pull/45354 )
</summary>
<hr>

	
<hr>
</details>

## Changelog
:cl:
fix: Fixes clients crashing when spammed with a lot of messages
/:cl:

Works in local, no runtimes and 100 roaches attacking together don't make me crash, but I suggest testmerging and checking if it works on live too first.

### CODE REVIEW

**Comment by TorinoFermic** [comment]
You have leave a working link for the tg pr to reference it

**Comment by Carbonhell** [comment]
Woops fixed

**Comment by PlinHost** [comment]
Do icons work?

**Comment by Carbonhell** [comment]
> 
> 
> Do icons work?

What type of icons? Stuff like emojis? Do you have an use case I can test?

**Comment by drexample** [comment]
@Carbonhell Have you tested it working with emojis? They are already in code.

**Comment by mattroks101** [comment]
@drexample it doesn't mess up emojis at all. I have tested this on my branch and even warfare it seems to work. 

However it makes the `x2` thing appear on the line below the message. It does this on TGMC as well. Does anyone know why it does this? I find it a bit distracting but if it stops crashes it is the preferable option.
dreamseeker_9ZGta7tGwG

**Comment by SpadesSlick** [comment]
image
Cyrillic is broken.

File: code/controllers/subsystems/chat.dm [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	if(!istext(message))
+		crash_with("to_chat called with invalid input type")
+		return
+
+	if(target == world)
+		target = clients
+
+	//Some macros remain in the string even after parsing and fuck up the eventual output
+	message = replacetext(message, "\improper", "")
+	message = replacetext(message, "\proper", "")
```
Comment (PlinHost) [inline code/controllers/subsystems/chat.dm]: This is already handled in `strip_improper` below.

File: code/controllers/subsystems/chat.dm [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	if(handle_whitespace)
+		message = replacetext(message, "\n", "<br>")
+		message = replacetext(message, "\t", "[FOURSPACES][FOURSPACES]")
+	message += "<br>"
```
Comment (PlinHost) [inline code/controllers/subsystems/chat.dm]: There is no such line in `to_chat_immediate` so I guess it is not needed.

**Pull request merged**